### PR TITLE
Issue 1 - Adding Snapshots support for failing tests

### PR DESCRIPTION
**Pull request by liamjdouglas**
Snapshots will be saved for any failing test.
- Snapshots are saved into the derived data folder for the current simulator.
- Snapshots are saved in the folder structure LayoutTestImages/{CLASS-NAME}/{METHOD-NAME}/
- Snapshots are saved in the format Width-{VIEW-WIDth}_Height-{VIEW-HEIGHT}_{HASH-OF-FAILING-DATA-DESCRIPTION}.png
- Snapshots are logged to a index.html file for each class which includes the failure reason, image and data that was passed to the view.
- The paths to the failing test index.html files are logged to the console at the end of the test bundle run.
- Snapshots can be disabled using the LYTConfig
- Snapshots can be limited per method using the LYTConfig

### CODE REVIEW

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import <XCTest/XCTest.h>
+
+@interface LYTLayoutFailingTestSnapshotRecorder : NSObject<XCTestObservation>
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: We should add docs to this class, the properties and the methods.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+#import "LYTLayoutFailingTestSnapshotRecorder.h"
+#import "LYTConfig.h"
+#import "LYTLayoutTestCase.h"
+
+void SimpleLog(NSString *format, ...) {
+    // Log to the console without timestamp that NSLog gives us
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Why don't we want the timestamp?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+@property (nonatomic, strong) UIView *viewUnderTest;
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: Should these be publicly settable? Not sure exactly what these are for though...so maybe my question will be answered soon.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h (hunk 1 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: I'll add these.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 2 above)
Comment (KotiJocki) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: The timestamp made the logging look like part of the output of the tests, whereas this is a summary of the tests run, similar to "Test session log:". The timestamp isn't needed and obscures the essential information.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h (hunk 3 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: They need to be set from the LYTLayoutTestCase to keep track of the currently failing view. When 
`- (void)recordFailureWithDescription:(NSString *)description inFile:(NSString *)filePath atLine:(NSUInteger)lineNumber expected:(BOOL)expected {`
is called we don't have a reference to the view and data that failed. Options are to leave as is, change it to a method to set them rather than having them exposed or save the view and data in the LYTLayoutTestCase and pass them into saveImageOfCurrentViewWithInvocation when we get a failure?

Comment (KotiJocki) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: Another option would be to expose them privately to LYTLayoutTestCase

Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: It seems a little strange to have the singleton hold onto state for a specific test (though can see how it's necessary).

In the future, people may want to use this without an LYTLayoutTestCase, so I think we should try to anticipate that? Maybe holding onto them inside the test case makes the most sense given this? So no-one else will ever overwrite them by mistake?

Thoughts?

Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: Yeh that makes sense and feels cleaner passing them in with the `saveImageOfCurrentView` method. I'll make the change and have LYTLayoutTestCase hold onto the current view and data.

**Comment by liamjdouglas** [comment]
Added docs and updated so the images are stored in the test case and passed into the render when they should be saved.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ Folder structure for images follows the pattern: {DERIVED_DATA}/LayoutTestImages/{CLASS_NAME}/index.html
+ Index.html file includes tables with failure description, image and the data that the image failed with.
+ 
+  \param invocationClass The class to start a new log for, class name is used in the directory structure to save the snapshots.
+ */
+- (void)startNewLogForClass:(Class)invocationClass;
+
+/**
+ Finishes the index.html log for the current invocation class by adding closing html tags.
+ Can be called safely without calling startNetLogForClass: as no index.html file for the class with exists so nothing will happen.
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: startNew

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ Renders the view passed as a png and saves it. The failure description, image and data are added to the index.html file for the current test class.
+ The invocation is used to create a folder for the invocations method name and the failing views for that invocation are saved within. Folder structure for invocation follows the pattern: {DERIVED_DATA}/LayoutTestImages/{CLASS_NAME}/{METHOD_NAME}/
+ Images sare saved with the file name format: Width-{VIEW_WIDTH}_Height-{VIEW_HEIGHT}_Data-{DATA_DESCRIPTION_HASH}.png
+ 
+ \param view The view to save a image of.
+ \param data The data that was used to poulate the view
+ \param invocation The invocation for the current test. When used from a XCTestCase subclass the invocation will include the name of the test method currently being run.
+ \param failureDescription The description of why the view that is being saved failed it's layout tests.
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.h]: super minor :):

its, not it's

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    if ([testCase isKindOfClass:[LYTLayoutTestCase class]]) {
+        NSString *pathForSnapshots = [self commonRootPathForInvocationClass:[testCase class]];
+        pathForSnapshots = [pathForSnapshots stringByAppendingString:@"/index.html"];
+        [self.failingTestsSnapshotFolders addObject:pathForSnapshots];
+    }
+}
+
+- (void)startNewLogForClass:(Class)invocationClass {
+    self.invocationClass = invocationClass;
+    [self deleteCurrentFailingSnapshots];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: This confused me for a while. It looks like you're deleting all the snapshots...but you're just deleting the ones for the new invocation. Same with the one below.

Thoughts on making this more functional?
as in: deleteCurrentFailingSnapshotsForInvocationClass:

This way things like the order of lines here don't matter. Obviously this works too though, just wondered what you think. (same for methods below).

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        [UIImagePNGRepresentation(viewImage) writeToFile:imagePath atomically:YES];
+        [self appendViewWithWidth:view.frame.size.width
+                           height:view.frame.size.height
+                         viewData:data
+      toLogWithFailureDescription:failureDescription
+                   withInvocation:invocation];
+    }
+}
+
+- (BOOL)shouldSaveImageOfViewatPath:(NSString *)imagePath withInvocation:(NSInvocation *)invocation {
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: capitalize At

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+- (void)startNewLogForClass:(Class)invocationClass {
+    self.invocationClass = invocationClass;
+    [self deleteCurrentFailingSnapshots];
+    [self createIndexHTMLFile];
+}
+
+- (void)saveImageOfView:(UIView *)view withData:(NSDictionary *)data fromInvocation:(NSInvocation *)invocation failureDescription:(NSString *)failureDescription {
+    NSString *imagePath = [self pathForImageWithWidth:view.frame.size.width height:view.frame.size.height data:data invocation:invocation];
+    if ([self shouldSaveImageOfViewatPath:imagePath withInvocation:invocation]) {
+        [self createDirectoryForInvocationIfNeeded:invocation];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Yeah, this reads much better to me compared to above.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        UIImage *viewImage = [self renderLayer:view.layer];
+        [UIImagePNGRepresentation(viewImage) writeToFile:imagePath atomically:YES];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Can this fail? I think writeToFile returns a BOOL? Possibly, we should log if this fails (almost never will)

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+                           height:view.frame.size.height
+                         viewData:data
+      toLogWithFailureDescription:failureDescription
+                   withInvocation:invocation];
+    }
+}
+
+- (BOOL)shouldSaveImageOfViewatPath:(NSString *)imagePath withInvocation:(NSInvocation *)invocation {
+    if ([[NSFileManager defaultManager] fileExistsAtPath:imagePath]) {
+        return NO;
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: when would this happen? include a comment?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    NSString *filePath = [self indexHTMLFilePath];
+    NSError *error;
+    
+    NSString *header = @"<HTML><HEAD></HEAD><BODY><TABLE style='width:100%'><TR><TH>Description</TH><TH>Image</TH><TH>Input Data</TH></TR>";
+    
+    [fileManager createDirectoryAtPath:classNameDirectory withIntermediateDirectories:YES attributes:nil error:&error];
+    [fileManager createFileAtPath:filePath contents:nil attributes:nil];
+    
+    // Write to the file
+    [header writeToFile:filePath atomically:YES
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: again here. Why pass in an error if we're not reading it?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+               encoding:NSUTF8StringEncoding error:&error];
+}
+
+- (void)finishLog {
+    NSString *filePath = [self indexHTMLFilePath];
+    NSString *footer = @"</TABLE></BODY></HTML>";
+    NSFileHandle *fileHandler = [NSFileHandle fileHandleForUpdatingAtPath:filePath];
+    
+    [fileHandler seekToEndOfFile];
+    [fileHandler writeData:[footer dataUsingEncoding:NSUTF8StringEncoding]];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Again here. Care about failures?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    [fileHandler closeFile];
+}
+
+- (void)appendViewWithWidth:(CGFloat)width
+                     height:(CGFloat)height viewData:(NSDictionary *)data
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: new line for viewData

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+- (void)deleteCurrentFailingSnapshots {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    [fileManager removeItemAtPath:[self commonRootPath] error:nil];
+}
+
+- (void)createDirectoryForInvocationIfNeeded:(NSInvocation *)invocation {
+    NSString *directoryPath = [self directoryPathForCurrentInvocation:invocation];
+    BOOL isDirectory = NO;
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: why do we need this? This isn't read. Can you not pass in nil?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+- (NSString *)methodNameForInvocation:(NSInvocation *)invocation {
+    return NSStringFromSelector((SEL)[invocation selector]);
+}
+
+- (NSString *)commonRootPath {
+    return [self commonRootPathForInvocationClass:self.invocationClass];
+}
+
+- (NSString *)commonRootPathForInvocationClass:(Class)classForRoot {
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Oh, you already have this method. I'd take out the helper method then and just manually pass in self.invocationClass. Slightly more typing, but clearer I think

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+ */
+- (NSString *)pathForImageWithWidth:(CGFloat)width height:(CGFloat)height data:(NSDictionary *)data invocation:(NSInvocation *)invocation {
+    NSString *directoryPath = [self directoryPathForCurrentInvocation:invocation];
+    NSString *imageName = [self nameForImageWithWidth:width height:height data:data];
+    return [directoryPath stringByAppendingPathComponent:imageName];
+}
+
+- (NSString *)nameForImageWithWidth:(CGFloat)width height:(CGFloat)height data:(NSDictionary *)data {
+    NSString *imageName = [NSString stringWithFormat:@"Width-%.2f_Height-%.2f_Data-%lu", width, height, (unsigned long)data.description.hash];
+    return [imageName stringByAppendingString:@".png"];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Random question...any reason you picked PNG over JPG? I suppose no loss is a good argument since you might spend some time staring at individual pixels?

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+- (UIImage *)renderLayer:(CALayer *)layer {
+    UIImage *snapshot = nil;
+        CGRect bounds = layer.bounds;
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: format

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    if (CGRectGetWidth(bounds) > 0 && CGRectGetHeight(bounds) > 0) {
+        UIGraphicsBeginImageContextWithOptions(bounds.size, NO, 0);
+        CGContextRef context = UIGraphicsGetCurrentContext();
+        
+        CGContextSaveGState(context);
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: I think we want to wrap this in an if (context) statement? I think this only happens if the bounds has no area, but maybe it can happen for otehr reasons too? Not sure if this will crash.

File: LayoutTest/TestCase/LYTLayoutTestCase.m [hunk 19]
```diff
@@ -29,6 +31,16 @@ @implementation LYTLayoutTestCase
 
 #pragma mark - Running Tests
 
++ (void)setUp {
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutTestCase.m]: It looks like you're overriding the class function here. I didn't know this even exists...I assume this is called once before all tests in this class are run? While the instance method gets run before each test?

File: LayoutTest/TestCase/LYTLayoutTestCase.m [hunk 20]
```diff
@@ -187,6 +205,15 @@ - (void)failTest:(NSString *)errorMessage view:(nullable UIView *)view {
     XCTFail(@"%@", errorMessage);
 }
 
+#pragma mark Failing Test Snapshots
+
+- (void)recordFailureWithDescription:(NSString *)description inFile:(NSString *)filePath atLine:(NSUInteger)lineNumber expected:(BOOL)expected {
+    [super recordFailureWithDescription:description inFile:filePath atLine:lineNumber expected:expected];
+    if (self.enableFailingTestSnapshots) {
+        [[LYTLayoutFailingTestSnapshotRecorder sharedInstance] saveImageOfView:self.viewUnderTest withData:self.dataForViewUnderTest fromInvocation:self.invocation failureDescription:description];
```
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutTestCase.m]: huh...never knew that invocation was a property on XCTestCase...very cool.

File: LayoutTestBase/Config/LYTConfig.h [hunk 21]
```diff
@@ -71,6 +71,15 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic) BOOL accessibilityTestsEnabled;
 
 /**
+ If on, a snapshot of the view for the current failing test will be saved to the derived data folder. The data for the failing test is also logged in the index.html file for each LYTLayoutTestCase sub class. The path to the folder is logged after the test suite has finished.
+ 
+ This property can also be turned on/off on each LayoutTestCase.
+ 
+ Default: true
+ */
+@property (nonatomic) BOOL enableFailingTestSnapshots;
```
Comment (plivesey) [inline LayoutTestBase/Config/LYTConfig.h]: Can we rename to match style of the other properties.

failingTestSnapshotsEnabled

also, maybe even:

testFailureSnapshotsEnabled?

File: LayoutTestBase/Config/LYTConfig.h [hunk 22]
```diff
@@ -107,6 +116,21 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (nonatomic) CGFloat cgFloatEpsilon;
 
+
+/**
+ Default snapshots to save per method.
+ */
+#define SaveUnlimitedSnapshotsPerMethod -1
```
Comment (plivesey) [inline LayoutTestBase/Config/LYTConfig.h]: #define doesn't work in swift projects. I think we need to make this an extern NSInteger like the constants in LYTDeviceConstants.h

Also, since it's in the global space, we should prefix with LYT

File: LayoutTestTests/ConfigTests.m [hunk 23]
```diff
@@ -51,8 +53,38 @@ - (void)testConfigDefaultLayoutTest {
     XCTAssertEqual(testCase.ambiguousAutolayoutTestsEnabled, [LYTConfig sharedInstance].ambiguousAutolayoutTestsEnabled);
     XCTAssertEqual(testCase.interceptsAutolayoutErrors, [LYTConfig sharedInstance].interceptsAutolayoutErrors);
     XCTAssertEqual(testCase.accessibilityTestsEnabled, [LYTConfig sharedInstance].accessibilityTestsEnabled);
+    XCTAssertEqual(testCase.enableFailingTestSnapshots, [LYTConfig sharedInstance].enableFailingTestSnapshots);
     XCTAssertEqual(testCase.viewClassesAllowingSubviewErrors, [LYTConfig sharedInstance].viewClassesAllowingSubviewErrors);
     XCTAssertEqual(testCase.viewClassesRequiringAccessibilityLabels, [LYTConfig sharedInstance].viewClassesRequiringAccessibilityLabels);
 }
 
+- (void)testResetDefaultsResetsToExpectedValues {
```
Comment (plivesey) [inline LayoutTestTests/ConfigTests.m]: nice addition

File: SampleApp/SampleAppTests/SampleFailingLayoutTestsWithSnapshots.swift [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+        runLayoutTests() { (view: SampleFailingView, data: [NSObject: AnyObject], context: Any?) in
+            // Expecting one of the auto-layout tests to fail when the really long string causes the label to overlap with the button
+        }
+    }
+}
+
+extension SampleFailingView : LYTViewProvider {
+    public class func dataSpecForTest() -> [NSObject: AnyObject] {
+        return [
+            "text": "A really, really, really, really, really, really, really, long string",
```
Comment (plivesey) [inline SampleApp/SampleAppTests/SampleFailingLayoutTestsWithSnapshots.swift]: If we make this LYTStringValues() too, won't this also fail? Because it'll test it on two long strings at some point?

Good idea to add this to the sample app. This way we'll show off more of what the library can do.

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+- (void)testSaveImageOfCurrentViewSavesImageToDisk {
+    UIView *testView = [[UIView alloc] initWithFrame:(CGRect){0, 0, 100, 100}];
+    [self.recorder startNewLogForClass:self.class];
+    
+    [self.recorder saveImageOfView:testView withData:@{@"key" : @"value"} fromInvocation:self.invocation failureDescription:@""];
+    
+    NSString *dataHash = @"";
+#ifdef __LP64__
```
Comment (plivesey) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: What is this? Can you add a comment?

Comment (plivesey) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: Instead of this, why not just call .description.hash on the data? That way it'll be future proof too?

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+    dataHash = @"7040999336831705875";
+#else
+    dataHash = @"2409456403";
+#endif
+    NSString*expectedImageName = [NSString stringWithFormat:@"/Width-100.00_Height-100.00_Data-%@.png", dataHash];
+    NSString *imagePath = [[self pathForCurrentInvocation] stringByAppendingString:expectedImageName];
+    XCTAssertTrue([[NSFileManager defaultManager] fileExistsAtPath:imagePath]);
+}
+
+- (void)testSaveImageOfCurrentViewWhenDataDescriptionIsOverTwoHunderedCharactersImageSavedWithExpectedName {
```
Comment (plivesey) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: whats special about two hundred characters? File name is too long?

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+- (NSString *)expectedEndOfFileHTML {
+    return @"</TABLE></BODY></HTML>";
+}
+
+- (NSString *)threeHunderedAndSixtyFourCharacterString {
+    return @"Very long string. This string is so long that it's longer than I want it to be. Just a really really long string. In fact, it's just long as long can be. I'm just lengthening the longest string by making it longer with more characters. Do you think this is long enough yet? I think so, so I'm going to stop making this long string longer by adding more characters.";
+}
+
```
Comment (plivesey) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: pretty epic test coverage :+1:

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 9 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Yeh it does return a BOOL. Could log the failure to the console but it will likely get lost in the test output. Other problem is that it'll still write to the index.html file but the image won't be there. In this case there is the option to only update the log if we successfully write to disk, otherwise update the log with a failure message. Happy to add that but like you say should almost never happen.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 10 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Added incase there is a view/data combo failure that is repeated. Unlikely to happen I think but should help save some time creating the image and writing it again.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 11 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Sorry out of habit, will remove it.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 2 above)
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: :+1:

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 6 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Updated the to pass in the class to delete image for.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 12 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Same as above, I'm not concerned about it but happy to add something if needed.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 14 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Sorry out of habit, will remove it.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 15 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Will remove it.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 16 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Yeh it was just so nothing was lost.

File: LayoutTest/TestCase/LYTLayoutTestCase.m (hunk 19 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutTestCase.m]: Yeh the class method version is called at the start of the test class being run before any tests are run. Similarly `+(void)tearDown;` is called after all the tests in the class have run.

File: LayoutTestBase/Config/LYTConfig.h (hunk 21 above)
Comment (liamjdouglas) [inline LayoutTestBase/Config/LYTConfig.h]: Renamed to failingTestSnapshotsEnabled

File: LayoutTestBase/Config/LYTConfig.h (hunk 22 above)
Comment (liamjdouglas) [inline LayoutTestBase/Config/LYTConfig.h]: Done

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m (hunk 25 above)
Comment (liamjdouglas) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: Much better idea, I'll update.

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m (hunk 26 above)
Comment (liamjdouglas) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: This is a hang over for when the data description was being added to the file name as OS X has a max file name length of 250. I think it's still a valid test though to cover the edge case.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 18 above)
Comment (liamjdouglas) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: Added to stop this being logged to the console multiple times.

> xctest[25457] <Error>: CGContextSaveGState: invalid context 0x0. If you want to see the backtrace, please set CG_CONTEXT_SHOW_BACKTRACE environmental variable.

Happy to remove as it's just dealing with a edge case.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 16 above)
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: :+1:

File: LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m (hunk 26 above)
Comment (plivesey) [inline LayoutTestTests/LayoutTestCaseTests/LYTLayoutFailingTestSnapshotRecorderTests.m]: Ah, got it. Yes, good to show that this is being handled.

File: LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m (hunk 18 above)
Comment (plivesey) [inline LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m]: No, don't remove. I've seen a lot of these when we run tests at LinkedIn...good to get rid of these logs.

**Comment by plivesey** [comment]
Ok, the code LGTM. Haven't yet pulled it down and ran it on my machine, so will do that now. In the meantime, do you want to merge the commits?

**Comment by liamjdouglas** [comment]
I've squashed the commits so whenever you are happy it's good to be merged.

**Comment by plivesey** [comment]
Are you sure? The title of this page still reads: "liamjdouglas  wants to merge 88 commits into linkedin:master from liamjdouglas:issue-1-adding-snapshots". Not sure how to verify the number of commits the merge is actually going to add?

**Comment by plivesey** [comment]
Btw, not needed for this commit, but idea for the future is to make the SampleApp a bit clearer in that views are overlapping. As in, if you saw the button behind the label, its much more obvious something has gone wrong. However, when I changed the background color of the label, the button was still hidden...

Anyway, just an idea for onboarding people onto the library, but not necessary for this commit.

**Comment by plivesey** [comment]
Ah...worked it out. There is actually a small bug in the sample app.

You should use: ```button.setTitle(json["buttonText"] as? String, forState: .Normal)``` instead of ```button.titleLabel.text =```. The second command actually always erases the title.

File: SampleApp/SampleApp/SampleFailingView.swift [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+    @IBOutlet weak var label: UILabel!
+    @IBOutlet weak var button: UIButton!
+    
+    class func loadFromNib() -> SampleFailingView {
+        return NSBundle.mainBundle().loadNibNamed(SampleFailingView.nibName, owner: nil, options: nil)[0] as! SampleFailingView
+    }
+
+    func setup(json: [NSObject: AnyObject]) {
+        label.text = json["text"] as? String
+        button.titleLabel?.text = json["buttonText"] as? String
```
Comment (plivesey) [inline SampleApp/SampleApp/SampleFailingView.swift]: Use button.setTitle(json["buttonText"] as? String, forState: .Normal) instead.

This code currently does nothing (I don't think that's intended). The above code produces much better screenshots.

**Comment by plivesey** [comment]
Great, looks good. Going to merge now.
That was a big commit. Thanks for all the hard work. :)

**Comment by plivesey** [comment]
Did a last check before merging...and it doesn't look like the commits have been squashed in your branch? I see:

```
commit c6055f4f56bd27e9bf63c8efcd89353e6d6b88da
    Issue #1 Add the ability to take snapshots of views (especially for failed tests)
    Fixing bug in sample app

commit 2c78c80f88c3039eba931bcf3a3f85e21c00b93c
    Merge branch 'issue-1-adding-snapshots' of github.com:liamjdouglas/LayoutTest-iOS into issue-1-adding-snapshots

commit d1d4cf384d58550fca815f5d497c72ce078f8b40
    Issue #1 Adding snapshots afor failing tests

commit 7e326ca01d5eac4cc030028b7df8d34679c44460
Merge: d931a9a 1fd5600

    Merge branch 'master' of github.com:liamjdouglas/LayoutTest-iOS into issue-1-adding-snapshots
etc.
```

Can you check to make sure that they are squashed?

**Comment by liamjdouglas** [comment]
Not sure what happen with the squash but I can't get the branch to drop the other 89 commits now. Made this with everything squashed https://github.com/linkedin/LayoutTest-iOS/pull/9.
